### PR TITLE
Add support for Config, endpoint_mode and 3.3 support

### PR DESCRIPTION
**Pull request by xianlubird**
```
Version 3.3
An upgrade of version 3 that introduces new parameters only available with Docker Engine version 17.06.0+, and higher.

Introduces the following additional parameters:

    build labels
    credential_spec
    configs
    deploy endpoint_mode
```

Docker compose v3.3 add four attributes.   For `credential_spec` we shouldn't support. I will work on it , support v3.3 ASAP.

Issue:  https://github.com/kubernetes/kompose/issues/914 and https://github.com/kubernetes/kompose/issues/865

@cdrage

### CODE REVIEW

**Comment by xianlubird** [comment]
@cdrage I send you a mail which title start with `charlie@charliedrage.com` address. Please check it if you are free. Thank you .

**Comment by xianlubird** [comment]
@cdrage @hangyan 
A brief introduction:
- This PR implement compose v3.3 support for kubernetes transform. For openshift, it will be implemented in another PR later.
- `credential_spec ` can only be used in windows container, so we ignore it 
- `configs ` convert to `configmap` in kubernetes and can be mount into deployment pod
- `endpoint_mode` vip mode is similar to `nodePort` type service, `dnsrr` should ignore

@cdrage @hangyan Pls review this PR if you are free. Review this PR maybe cost much time. If you have any comments , I will refactor my code.  Thanks . :-)

**Comment by youyajie** [comment]
What's update on this ?

**Comment by xianlubird** [comment]
@youyajie Thank you for notice this pr.  @cdrage @hangyan is reviewing code , hope it will be merged soon :-)

**Comment by hangyan** [comment]
@cdrage Sure. I am current working on the secret support, separate the commit can also reduce the potentially conflict

File: script/test/fixtures/compose-v3.3-test/compose-endpoint-mode-1.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+services:
+  wordpress:
+    image: wordpress
+    ports:
+      - "8080:80"
+    deploy:
+      mode: replicated
+      replicas: 2
+      endpoint_mode: vip
```
[the remark this answers is not in the document]
Comment (xianlubird) [reply script/test/fixtures/compose-v3.3-test/compose-endpoint-mode-1.yaml]: Done  @cdrage

**Comment by xianlubird** [comment]
@cdrage Any update ?

File: docs/conversion.md [hunk 2]
```diff
@@ -4,7 +4,7 @@ This document outlines all possible conversion details regarding `docker-compose
 
 The current table covers all **current** possible Docker Compose keys.
 
-__Note:__ due to the fast-pace nature of Docker Compose version revisions, minor versions such as 2.1, 2.2, 3.3 are not supported until they are cut into a major version release such as 2 or 3.
+__Note:__ due to the fast-pace nature of Docker Compose version revisions, minor versions such as 2.1, 2.2 are not supported until they are cut into a major version release such as 2 or 3.
```
Comment (cdrage) [inline docs/conversion.md]: Go ahead and completely remove this line.. We should simply state that we don't support anything 3.4 and above at the moment.

File: pkg/loader/compose/v3.go [hunk 3]
```diff
@@ -572,3 +581,27 @@ func mergeComposeObject(oldCompose *types.Config, newCompose *types.Config) (*ty
...
+		return []string{}
+	}
+
+	var keysFound []string
+	for _, service := range composeObject.Services {
+		if service.CredentialSpec.Registry != "" || service.CredentialSpec.File != "" {
+			keysFound = append(keysFound, "credential_spec")
+		}
+		if service.Deploy.EndpointMode == "dnsrr" {
+			keysFound = append(keysFound, "dnsrr")
```
[the remark this answers is not in the document]
Comment (cdrage) [reply pkg/loader/compose/v3.go]: ah, that makes sense. thank you.

**Comment by hangyan** [comment]
@cdrage @xianlubird 

A few thoughts:

1. OpenShfit support should be added anytime
2. About deploy mode. In my opinion, vip -> ClusterIP , dnsrr-> Headless

**Comment by hangyan** [comment]
Also, the current implement of ConfigMap and Volumes Mount seems not correct. Please consider the following rules:

1. If target is optional, config name is `example`. the target path should be `/example`, so volumeMounts.mountPath = /  and volumeMounts.name=example and ConfigMap.Data=example:<data>

2. If target is set, say it's `/test`., config name is `example`. then volumeMounts.mountPath = / and volumeMounts.name=example and ConfigMap.Data=example:<data> and volumes. configMap. items[0].key = example and volumes. configMap. items[0].path=test

Please check this doc: https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#add-configmap-data-to-a-volume

@cdrage @xianlubird

**Comment by xianlubird** [comment]
@hangyan 

1. OpenShfit will be implemented in another pr, this pr focus on kubernetes
2. `VIP` means user can visit any `ip:port` in the cluster node, so it equals to `NodePort` service.  ClusterIP can't be visited outside the cluster. `dnsrr` means user can use service name to connect, so in k8s, a normal service type can connect through service name.

**Comment by hangyan** [comment]
@xianlubird 

1. OpenShift support is just a few commands, it would't take many time. I don't think it's a good idea to seperate this

2. According to the compose doc. `vip` mode acts like a normal load balancer, the client wouldn't known what the backends are. With NodePort, you have to known how many nodes are in the kubernetes  cluster, and what are there ips, I think these principle take preceder over the `access from outside of the cluster` rule.

`dnsrr` works very similar to Headless Service. In Headless service, when you do a dns-lookup for the service name, a range of ips will returned(the ip list of the pods).

**Comment by hangyan** [comment]
@cdrage Please take a look at my reviews and share some of your thoughts on this

**Comment by xianlubird** [comment]
@hangyan 

1.  For `vip` mode, I don't agree with you.   In `vip` mode, user should use `ip:port` to visit service. User also should know node ip, it totally equals to `NodePort` service.   If convert to `Cluster IP`, how should user to visit service outside cluster?  Cloud they use node ip with port ?  It can not equals to `ClusterIP` type.

`dnsrr` means user visit `service name`, `headless` is a external service. I think `cluster ip` is better. Both of them can visit through `service name`, but `cluster ip` type is more used.

2. I am not similar to OpenShift. So this pr first name is `add support 3.3 for kubernetes`. It will cost much time for me to implement OpenShift. And for OpenShift, I think it should split to another pr, it will more clear.


Need your opinion @cdrage

**Comment by xianlubird** [comment]
@cdrage @hangyan Discuss with @hangyan , I change some code for this pr:

- Ignore short syntax config. Because short syntax config mount path in `Deployment` must be `/`, but docker daemon don't all mount path is `/`, so we have to ignore it.
- For long syntax config, if target path is `/`, ignore this. The reason is similar to above.
- Add openshift support, thanks @hangyan insist
- Change document according to @cdrage advice

Now this PR has become a `XXL` size, very exciting to see that it will be merged.

Thanks all for your review.

**Comment by xianlubird** [comment]
@hangyan any update on this ?

**Comment by xianlubird** [comment]
@cdrage 
We still have some differences between @hangyan at which service type we should convert.

- For compose `network_mode: vip`, according to docker's document [link](https://docs.docker.com/engine/swarm/ingress/#publish-a-port-for-a-service), `vip` mode is the `swarm mode routing mesh`, equals to `docker service --publish`. So user can visit any node ip in the cluster, the address is `ip:port`.  It is especially like `NodePort` service in kubernetes cluster which user can visit any node ip in the cluster, the address is also `ip:port`.  

- For compose `network_mode: dnsrr`. This allow user to visit `service name` in a container in swarm mode cluster.  It's similar to `Cluster IP` type in kubernetes cluster. For `Headless` service type, it is very similar to compose attribute [external_links](https://docs.docker.com/compose/compose-file/#external_links).

Now we need @cdrage opinion to push this pr going on. Hope you have time to see this and give us some advice. So we can merge this important pr soon. :-)

Many thanks.

**Comment by cdrage** [comment]
@xianlubird For NodePort it's actually the user can visit any node *port*. You still have to figure out the IP. network_mode: vip in my opinion is closer to LoadBalancer.

I agree however with `dnsrr`. 

I will do another review in the next few days!

**Comment by xianlubird** [comment]
@cdrage any update on this ?

**Comment by hangyan** [comment]
@xianlubird @cdrage I have re-check the kubernetes documents, it seems like subPath(https://kubernetes.io/docs/concepts/storage/volumes/#using-subpath) can bring the full support for compose `config`.

**Comment by fentas** [comment]
@hangyan any progress?

**Comment by cdrage** [comment]
@hangyan @xianlubird Minor conflicting file, otherwise this LGTM. I can do another review after the conflicted file is resolved!

**Comment by xianlubird** [comment]
@cdrage All check have passed.   conflicting solved.

**Comment by hangyan** [comment]
@xianlubird As i have point out before, subPath can bring the full support for compose config, otherwise this PR is too limited. @cdrage Or we can merge this first and create a new one to improve it

**Comment by cdrage** [comment]
@hangyan Adding subPath may be best before merging, so that `config` support is fully supported (short-syntax and all).

Although to be honest, 3.3 support seems to be important as well as endpoint_mode. Maybe we should merge and then open an issue indicating that we should implement subPath / syntax support for config.

**Comment by cdrage** [comment]
@hangyan I agree. It's been way too long. Any issues that we encounter in the future can be resolved.

Merging this *massive* PR.

Special thanks to @xianlubird ! :100:

**Pull request merged**